src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -15,42 +14,74 @@ namespace System.Formats.Tar
     public abstract partial class TarEntry
     {
         internal TarHeader _header;
+
         // Used to access the data section of this entry in an unseekable file
         private TarReader? _readerOfOrigin;
 
-        // Constructor used when reading an existing archive.
-        internal TarEntry(TarHeader header, TarReader readerOfOrigin)
+        // Constructor called when reading a TarEntry from a TarReader.
+        internal TarEntry(TarHeader header, TarReader readerOfOrigin, TarEntryFormat format)
         {
+            Debug.Assert(header._format == format);
             _header = header;
+            _header._format = format;
             _readerOfOrigin = readerOfOrigin;
         }
 
-        // Constructor called when creating a new 'TarEntry*' instance that can be passed to a TarWriter.
+        // Constructor called when creating a new TarEntry.
         internal TarEntry(TarEntryType entryType, string entryName, TarEntryFormat format)
         {
             ArgumentException.ThrowIfNullOrEmpty(entryName);
-
-            // Throws if format is unknown or out of range
-            TarHelpers.VerifyEntryTypeIsSupported(entryType, format, forWriting: false);
-
-            _readerOfOrigin = null;
+            TarHelpers.ThrowIfEntryTypeNotSupported(entryType, format);
 
             _header = default;
+            _header._format = format;
 
-            _header._extendedAttributes = new Dictionary<string, string>();
-
+            // Default values for fields shared by all supported formats
             _header._name = entryName;
-            _header._linkName = string.Empty;
-            _header._typeFlag = entryType;
             _header._mode = (int)TarHelpers.DefaultMode;
+            _header._uid = 0;
+            _header._gid = 0;
+            _header._size = 0;
```

Comment:
The explicit assignments of uid/gid/size feel unnecessary to me, given that it's following `_header = default`.